### ISSUE TITLE
ICE on nightly-2022-03-18 when compilation ordinary Rust project.

### ISSUE THREAD

**Issue by vertexclique** [rust-lang/rust#95623]
### Code

```Rust
async-process
```

### Meta

`rustc --version --verbose`:
```
rustc 1.61.0-nightly (1bfe40d11 2022-03-18)
binary: rustc
commit-hash: 1bfe40d11c3630254504fb73eeccfca28d50df52
commit-date: 2022-03-18
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Error output

```
   Compiling async-process v1.3.0
thread 'rustc' panicked at 'assertion failed: !value.has_escaping_bound_vars()', /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/compiler/rustc_middle/src/ty/sty.rs:1089:9
stack backtrace:
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'assertion failed: !value.has_escaping_bound_vars()', /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/compiler/rustc_middle/src/ty/sty.rs:1089:9
stack backtrace:
   0:     0x7ff6afe5c96d - std::backtrace_rs::backtrace::libunwind::trace::ha3a6b21151c60a7f
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ff6afe5c96d - std::backtrace_rs::backtrace::trace_unsynchronized::hef2c3337aca2e2b3
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ff6afe5c96d - std::sys_common::backtrace::_print_fmt::h7100e2252d44bdd1
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7ff6afe5c96d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h388ab420cf9330db
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7ff6afeb644c - core::fmt::write::hb92282007cec191f
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/core/src/fmt/mod.rs:1190:17
   5:     0x7ff6afe4def1 - std::io::Write::write_fmt::hd6ade7727f8519fc
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/io/mod.rs:1655:15
   6:     0x7ff6afe5f9e5 - std::sys_common::backtrace::_print::he96f7ed8f0451fec
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7ff6afe5f9e5 - std::sys_common::backtrace::print::h827c968ad14125ac
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7ff6afe5f9e5 - std::panicking::default_hook::{{closure}}::hfcd1f9d84623f11f
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/panicking.rs:295:22
   9:     0x7ff6afe5f699 - std::panicking::default_hook::h7d1fe18b8808947c
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/panicking.rs:314:9
  10:     0x7ff6b05f9721 - rustc_driver[6650adb4e8a07042]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7ff6afe60130 - std::panicking::rust_panic_with_hook::haf4da044897eaa77
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/panicking.rs:702:17
  12:     0x7ff6afe5ff29 - std::panicking::begin_panic_handler::{{closure}}::h67c8a4c9c24c3542
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/panicking.rs:586:13
  13:     0x7ff6afe5ce24 - std::sys_common::backtrace::__rust_end_short_backtrace::h3f20aba2192957d4
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7ff6afe5fc99 - rust_begin_unwind
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/panicking.rs:584:5
  15:     0x7ff6afe23c23 - core::panicking::panic_fmt::hf49e1497732e8ea4
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/core/src/panicking.rs:143:14
  16:     0x7ff6afe23aed - core::panicking::panic::ha88b33e797e5f889
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/core/src/panicking.rs:48:5
  17:     0x7ff6b222a35f - rustc_trait_selection[de84fd39e101a068]::traits::type_known_to_meet_bound_modulo_regions
  18:     0x7ff6b1ac41c5 - <rustc_infer[e7c0dc8a8349484a]::infer::InferCtxtBuilder>::enter::<bool, rustc_ty_utils[1c2b436d223d6e4e]::common_traits::is_item_raw::{closure#0}>
  19:     0x7ff6b2779fae - rustc_ty_utils[1c2b436d223d6e4e]::common_traits::is_item_raw
  20:     0x7ff6b1ea488c - rustc_query_system[319694dcf5dfa03c]::query::plumbing::try_execute_query::<rustc_query_impl[2480959737b7d51e]::plumbing::QueryCtxt, rustc_query_system[319694dcf5dfa03c]::query::caches::DefaultCache<rustc_middle[5f63c5fd1ac72588]::ty::ParamEnvAnd<rustc_middle[5f63c5fd1ac72588]::ty::Ty>, bool>>
  21:     0x7ff6b296f8fb - rustc_query_system[319694dcf5dfa03c]::query::plumbing::get_query::<rustc_query_impl[2480959737b7d51e]::queries::is_sized_raw, rustc_query_impl[2480959737b7d51e]::plumbing::QueryCtxt>
  22:     0x7ff6b23d42e5 - <rustc_middle[5f63c5fd1ac72588]::ty::Ty>::is_sized
  23:     0x7ff6b235e2fe - <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached
  24:     0x7ff6b2[375](https://github.com/vertexclique/callysto/runs/5807959075?check_suite_focus=true#step:7:375)276 - rustc_middle[5f63c5fd1ac72588]::ty::layout::layout_of
  25:     0x7ff6b1f04bfa - rustc_query_system[319694dcf5dfa03c]::query::plumbing::get_query::<rustc_query_impl[2480959737b7d51e]::queries::layout_of, rustc_query_impl[2480959737b7d51e]::plumbing::QueryCtxt>
  26:     0x7ff6b1fa1f30 - <rustc_query_impl[2480959737b7d51e]::Queries as rustc_middle[5f63c5fd1ac72588]::ty::query::QueryEngine>::layout_of
  27:     0x7ff6b2[380](https://github.com/vertexclique/callysto/runs/5807959075?check_suite_focus=true#step:7:380)778 - <alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>> as alloc[36d2302c246c3597]::vec::spec_from_iter::SpecFromIter<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>, core[d9a3b528d00e4bcd]::iter::adapters::GenericShunt<core[d9a3b528d00e4bcd]::iter::adapters::map::Map<core[d9a3b528d00e4bcd]::slice::iter::Iter<rustc_middle[5f63c5fd1ac72588]::ty::FieldDef>, <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached::{closure#5}::{closure#0}>, core[d9a3b528d00e4bcd]::result::Result<core[d9a3b528d00e4bcd]::convert::Infallible, rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutError>>>>::from_iter
  28:     0x7ff6b237fc1c - <alloc[36d2302c246c3597]::vec::Vec<alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>>> as alloc[36d2302c246c3597]::vec::spec_from_iter::SpecFromIter<alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>>, core[d9a3b528d00e4bcd]::iter::adapters::GenericShunt<core[d9a3b528d00e4bcd]::iter::adapters::map::Map<core[d9a3b528d00e4bcd]::slice::iter::Iter<rustc_middle[5f63c5fd1ac72588]::ty::VariantDef>, <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached::{closure#5}>, core[d9a3b528d00e4bcd]::result::Result<core[d9a3b528d00e4bcd]::convert::Infallible, rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutError>>>>::from_iter
  29:     0x7ff6b235def0 - <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached
  30:     0x7ff6b2375276 - rustc_middle[5f63c5fd1ac72588]::ty::layout::layout_of
  31:     0x7ff6b1f04bfa - rustc_query_system[319694dcf5dfa03c]::query::plumbing::get_query::<rustc_query_impl[2480959737b7d51e]::queries::layout_of, rustc_query_impl[2480959737b7d51e]::plumbing::QueryCtxt>
  32:     0x7ff6b1fa1f30 - <rustc_query_impl[2480959737b7d51e]::Queries as rustc_middle[5f63c5fd1ac72588]::ty::query::QueryEngine>::layout_of
   Compiling async-std v1.11.0
  33:     0x7ff6b2380778 - <alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>> as alloc[36d2302c246c3597]::vec::spec_from_iter::SpecFromIter<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>, core[d9a3b528d00e4bcd]::iter::adapters::GenericShunt<core[d9a3b528d00e4bcd]::iter::adapters::map::Map<core[d9a3b528d00e4bcd]::slice::iter::Iter<rustc_middle[5f63c5fd1ac72588]::ty::FieldDef>, <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached::{closure#5}::{closure#0}>, core[d9a3b528d00e4bcd]::result::Result<core[d9a3b528d00e4bcd]::convert::Infallible, rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutError>>>>::from_iter
  34:     0x7ff6b237fc1c - <alloc[36d2302c246c3597]::vec::Vec<alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>>> as alloc[36d2302c246c3597]::vec::spec_from_iter::SpecFromIter<alloc[36d2302c246c3597]::vec::Vec<rustc_target[5d4685c11228a5f9]::abi::TyAndLayout<rustc_middle[5f63c5fd1ac72588]::ty::Ty>>, core[d9a3b528d00e4bcd]::iter::adapters::GenericShunt<core[d9a3b528d00e4bcd]::iter::adapters::map::Map<core[d9a3b528d00e4bcd]::slice::iter::Iter<rustc_middle[5f63c5fd1ac72588]::ty::VariantDef>, <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached::{closure#5}>, core[d9a3b528d00e4bcd]::result::Result<core[d9a3b528d00e4bcd]::convert::Infallible, rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutError>>>>::from_iter
  35:     0x7ff6b235def0 - <rustc_middle[5f63c5fd1ac72588]::ty::layout::LayoutCx<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt>>::layout_of_uncached
  36:     0x7ff6b2375276 - rustc_middle[5f63c5fd1ac72588]::ty::layout::layout_of
  37:     0x7ff6b1f04bfa - rustc_query_system[319694dcf5dfa03c]::query::plumbing::get_query::<rustc_query_impl[2480959737b7d51e]::queries::layout_of, rustc_query_impl[2480959737b7d51e]::plumbing::QueryCtxt>
  38:     0x7ff6b1fa1f30 - <rustc_query_impl[2480959737b7d51e]::Queries as rustc_middle[5f63c5fd1ac72588]::ty::query::QueryEngine>::layout_of
  39:     0x7ff6b1fa87e5 - rustc_codegen_ssa[82efcbacec23bfcf]::debuginfo::type_names::push_debuginfo_type_name
  40:     0x7ff6b1fa7f66 - rustc_codegen_ssa[82efcbacec23bfcf]::debuginfo::type_names::push_debuginfo_type_name
  41:     0x7ff6b1fa6770 - rustc_codegen_ssa[82efcbacec23bfcf]::debuginfo::type_names::compute_debuginfo_type_name
  42:     0x7ff6b18543bc - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::type_di_node
  43:     0x7ff6b185234c - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::build_pointer_or_reference_di_node
  44:     0x7ff6b1853805 - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::type_di_node
  45:     0x7ff6b184f419 - <&mut rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::build_struct_type_di_node::{closure#0}::{closure#0} as core[d9a3b528d00e4bcd]::ops::function::FnOnce<((usize, &rustc_middle[5f63c5fd1ac72588]::ty::FieldDef),)>>::call_once
  46:     0x7ff6b1814f3b - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::type_map::build_type_with_children::<rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::build_struct_type_di_node::{closure#0}, rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::build_struct_type_di_node::{closure#1}>
  47:     0x7ff6b1853545 - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::type_di_node
  48:     0x7ff6b185234c - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::build_pointer_or_reference_di_node
  49:     0x7ff6b1853805 - rustc_codegen_llvm[3b78a86e98ab98ba]::debuginfo::metadata::type_di_node
  50:     0x7ff6b1834e58 - <rustc_codegen_llvm[3b78a86e98ab98ba]::context::CodegenCx as rustc_codegen_ssa[82efcbacec23bfcf]::traits::debuginfo::DebugInfoMethods>::dbg_scope_fn
  51:     0x7ff6b1860dc2 - rustc_codegen_ssa[82efcbacec23bfcf]::mir::codegen_mir::<rustc_codegen_llvm[3b78a86e98ab98ba]::builder::Builder>
  52:     0x7ff6b181f3d4 - rustc_codegen_llvm[3b78a86e98ab98ba]::base::compile_codegen_unit::module_codegen
  53:     0x7ff6b24ddd36 - <rustc_query_system[319694dcf5dfa03c]::dep_graph::graph::DepGraph<rustc_middle[5f63c5fd1ac72588]::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle[5f63c5fd1ac72588]::ty::context::TyCtxt, rustc_span[3354b82f0c75eb5d]::symbol::Symbol, rustc_codegen_ssa[82efcbacec23bfcf]::ModuleCodegen<rustc_codegen_llvm[3b78a86e98ab98ba]::ModuleLlvm>>
  54:     0x7ff6b24cc969 - rustc_codegen_llvm[3b78a86e98ab98ba]::base::compile_codegen_unit
  55:     0x7ff6b24bab8c - <rustc_codegen_llvm[3b78a86e98ab98ba]::LlvmCodegenBackend as rustc_codegen_ssa[82efcbacec23bfcf]::traits::backend::CodegenBackend>::codegen_crate
  56:     0x7ff6b249bec7 - <rustc_session[77248bc14774052c]::session::Session>::time::<alloc[36d2302c246c3597]::boxed::Box<dyn core[d9a3b528d00e4bcd]::any::Any>, rustc_interface[1d66963a510d2f71]::passes::start_codegen::{closure#0}>
  57:     0x7ff6b248a938 - <rustc_interface[1d66963a510d2f71]::passes::QueryContext>::enter::<<rustc_interface[1d66963a510d2f71]::queries::Queries>::ongoing_codegen::{closure#0}::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<alloc[36d2302c246c3597]::boxed::Box<dyn core[d9a3b528d00e4bcd]::any::Any>, rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>
  58:     0x7ff6b24830bf - <rustc_interface[1d66963a510d2f71]::queries::Queries>::ongoing_codegen
  59:     0x7ff6b2446ebb - <rustc_interface[1d66963a510d2f71]::interface::Compiler>::enter::<rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}::{closure#2}, core[d9a3b528d00e4bcd]::result::Result<core[d9a3b528d00e4bcd]::option::Option<rustc_interface[1d66963a510d2f71]::queries::Linker>, rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>
  60:     0x7ff6b2459f2f - rustc_span[3354b82f0c75eb5d]::with_source_map::<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_interface[1d66963a510d2f71]::interface::create_compiler_and_run<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}>::{closure#1}>
  61:     0x7ff6b2447b04 - rustc_interface[1d66963a510d2f71]::interface::create_compiler_and_run::<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}>
  62:     0x7ff6b24448b2 - <scoped_tls[11b42f00ec6f76c4]::ScopedKey<rustc_span[3354b82f0c75eb5d]::SessionGlobals>>::set::<rustc_interface[1d66963a510d2f71]::interface::run_compiler<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}>::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>
  63:     0x7ff6b2442d5f - std[dc1d017f[401](https://github.com/vertexclique/callysto/runs/5807959075?check_suite_focus=true#step:7:401)af48a]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[1d66963a510d2f71]::util::run_in_thread_pool_with_globals<rustc_interface[1d66963a510d2f71]::interface::run_compiler<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}>::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>
  64:     0x7ff6b245aeb2 - <<std[dc1d017f401af48a]::thread::Builder>::spawn_unchecked_<rustc_interface[1d66963a510d2f71]::util::run_in_thread_pool_with_globals<rustc_interface[1d66963a510d2f71]::interface::run_compiler<core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>, rustc_driver[6650adb4e8a07042]::run_compiler::{closure#1}>::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>::{closure#0}, core[d9a3b528d00e4bcd]::result::Result<(), rustc_errors[e415a2b41ae93448]::ErrorGuaranteed>>::{closure#1} as core[d9a3b528d00e4bcd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  65:     0x7ff6afe6a2c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h[410](https://github.com/vertexclique/callysto/runs/5807959075?check_suite_focus=true#step:7:410)ea22b43d61d56
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/alloc/src/boxed.rs:1853:9
  66:     0x7ff6afe6a2c3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hff190b74d8f3d909
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/alloc/src/boxed.rs:1853:9
  67:     0x7ff6afe6a2c3 - std::sys::unix::thread::Thread::new::thread_start::h8cb9e648ae9f2254
                               at /rustc/58f11791af4f97572e7afd83f11cffe04bbbd12f/library/std/src/sys/unix/thread.rs:108:17
  68:     0x7ff6afd90609 - start_thread
  69:     0x7ff6afcaa293 - clone
  70:                0x0 - <unknown>


query stack during panic:
#0 [is_sized_raw] computing whether `dyn std::io::Write` is `Sized`
#1 [layout_of] computing layout of `*const dyn std::io::Write`
#2 [layout_of] computing layout of `core::ptr::unique::Unique<dyn std::io::Write>`
#3 [layout_of] computing layout of `alloc::boxed::Box<dyn std::io::Write>`
end of query stack
```

</p>
</details>

### Extra
Full log here:
https://github.com/vertexclique/callysto/runs/5807959075?check_suite_focus=true

**Comment by compiler-errors** [comment]
I believe this was fixed by #95020 -- @vertexclique, can you upgrade to the latest nightly and double check?

**Comment by vertexclique** [comment]
Works on the latest nightly. Cool, thanks!
